### ISSUE TITLE
Benchmarking HSTs with KDD99

### ISSUE THREAD

**Issue by nicholasjin** [creme-ml/creme#260]
Hi Max,

The documentation for HalfSpaceTrees claims

> They do not work well if anomalies are packed together in windows of time, such as in the KDD’99 HTTP and SMTP datasets.

I was wondering where this assertion comes from? In the original paper, the authors claim a .996 ROCAUC on the KDD99 HTTP dataset.

But when I ran HST on the dataset provided by `creme.datasets.fetchh_kdd99_http`, I got abysmal results (ROCAUC=0.5046). I am trying to understand the gap between the paper authors' results .

For reference, the code I ran was
```python
from creme import datasets
from creme import anomaly
import time
from sklearn.metrics import roc_auc_score

counter = 0
hst = anomaly.HalfSpaceTrees(
    n_trees=25,
    tree_height=15,
    window_size=250,
    scale=True
)

y_true = []
y_pred = []
start_time = time.time()
for x, y in datasets.fetch_kdd99_http():
    if counter%10000 == 0:
        print(f"iteration: {counter}; time elapsed: {time.time()-start_time}")
    if counter > 250:
        y_true.append(y)
        y_pred.append(hst.score_one(x))
    hst.fit_one(x)
    counter += 1
```

**Comment by MaxHalford** [comment]
Hey! There are a few reasons I can think of that are the cause of such a bad performance. I'm going to look into this closely. You can expect an answer very soon; I want to get this fixed for the next release.

**Comment by nicholasjin** [comment]
If there's anything I can do to help, let me know!

**Comment by MaxHalford** [comment]
Actually if you have some time on your hands you can try out [scikit-multiflow's implementation](https://scikit-multiflow.github.io/scikit-multiflow/_autosummary/skmultiflow.anomaly_detection.HalfSpaceTrees.html#skmultiflow.anomaly_detection.HalfSpaceTrees) and tell me how it goes; that will save me some time :)

**Comment by nicholasjin** [comment]
I achieved similar results with scikit-multiflow. Runtime: 21m, ROCAUC: 0.504
Note that skmultiflow does not implement a streaming StandardScaler, so that will skew results.

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from skmultiflow.data import DataStream
from skmultiflow.anomaly_detection import HalfSpaceTrees
from skmultiflow.evaluation import EvaluatePrequential

import h5py
import time

# Read in the http data
arrays = {}
f = h5py.File("http.mat")
for k,v in f.items():
    arrays[k] = np.array(v)

X = pd.DataFrame(arrays["X"]).T
y = pd.DataFrame(arrays["y"]).astype(int).T

stream = DataStream(data=X, y=y)
stream.prepare_for_use()

hst = HalfSpaceTrees(n_features=3)

counter = 0
y_true = []
y_pred = []
start_time = time.time()
while stream.has_more_samples():
    if counter % 10000 == 0:
        print(f"iteration: {counter}; time elapsed: {time.time()-start_time}")
    X, y = stream.next_sample()
    if counter > 250:
        yhat = hst.predict(X)
        y_pred.append(yhat)
        y_true.append(y)
    hst.partial_fit(X, y)
    counter += 1

from sklearn.metrics import roc_auc_score
roc_auc_score(y_true, y_pred)
```

**Comment by MaxHalford** [comment]
Thanks! To be honest I don't how the authors got to those results in their paper... I'm starting to think that they did something wrong in their experiments or "cheated" in some way. The HTTP dataset contains bursts of anomalies and therefore HST shouldn't work well on it. Shuffling the dataset helps a lot but isn't realistic. Also I assume they're prescaling their dataset and not scaling online, which also isn't realistic.

I've pushed some changes that I think improve the method. I've ran some sanity checks on some dummy datasets and it looks good to me. Notably I've changed the way the weighting works; instead of using the terminal leaf I include the weights from all the nodes along the way to the leaf. Also, instead of using the tree construction scheme in the paper I've implemented a custom tree construction algorithm which gives better performance. It's pretty simple and is actually something I'm using for a personal research project. I call it "padded trees".

I've also added the SMTP dataset to `creme`.
